File: spider.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import os
from dataclasses import dataclass
from enum import Enum
from collections import OrderedDict

import asyncio
import aiohttp
from tabulate import tabulate
from bs4 import BeautifulSoup
from aiohttp import ClientSession

MAX_MINUTES = 999

stations = [
    {"iStationId": 158789, "vStationName": "26栋1号机", "area": "四组团-26栋"},
    {"iStationId": 158742, "vStationName": "26栋2号机", "area": "四组团-26栋"},
    {"iStationId": 158664, "vStationName": "26栋3号机", "area": "四组团-26栋"},
    {"iStationId": 158507, "vStationName": "27栋1号机", "area": "四组团-27栋"},
    {"iStationId": 158748, "vStationName": "27栋2号机", "area": "四组团-27栋"},
    {"iStationId": 158501, "vStationName": "27栋3号机", "area": "四组团-27栋"},
    {"iStationId": 158764, "vStationName": "27栋4号机", "area": "四组团-27栋"},
    {"iStationId": 158666, "vStationName": "27栋5号机", "area": "四组团-27栋"},
    {"iStationId": 158661, "vStationName": "27栋6号机", "area": "四组团-27栋"},
    {"iStationId": 158668, "vStationName": "27栋7号机", "area": "四组团-27栋"},
    {"iStationId": 117387, "vStationName": "24栋1号机", "area": "四组团-24栋"},
    {"iStationId": 117474, "vStationName": "24栋2号机", "area": "四组团-24栋"},
    {"iStationId": 117374, "vStationName": "17栋（8插座）1号机", "area": "四组团-17栋"},
    {"iStationId": 117372, "vStationName": "17栋（8插座）2号机", "area": "四组团-17栋"},
    {"iStationId": 165088, "vStationName": "17栋（8插座）3号机", "area": "四组团-17栋"},
    {"iStationId": 158676, "vStationName": "17栋（10插座）1号机", "area": "四组团-17栋"},
    {"iStationId": 157805, "vStationName": "17栋（10插座）2号机", "area": "四组团-17栋"},
    {"iStationId": 157551, "vStationName": "17栋（10插座）3号机", "area": "四组团-17栋"},
    # {"iStationId": 117379, "vStationName": "第十餐厅1号机", "area": "三组团"},
    # {"iStationId": 117392, "vStationName": "第十餐厅2号机", "area": "三组团"},
    # {"iStationId": 117377, "vStationName": "游泳馆2号机", "area": "二组团"},
    # {"iStationId": 117381, "vStationName": "游泳馆1号机", "area": "二组团"},
    # {"iStationId": 117369, "vStationName": "5栋2号机", "area": "一组团"},
    # {"iStationId": 117375, "vStationName": "5栋1号机", "area": "一组团"},
]


@dataclass
class ChargingStation:
    class Status(Enum):
        USING = "使用中"
        AVAILABLE = "空闲中"
        UNAVAILABLE = "维护中"

    stationName: str
    outletName: str
    area: str
    status: Status = Status.AVAILABLE
    power: int = 0
    usedMinutes: int = 0
    totalMinutes: int = 0

    @property
    def remainingMinutes(self) -> int:
        return (self.totalMinutes - self.usedMinutes
                if self.status != ChargingStation.Status.UNAVAILABLE else MAX_MINUTES)

    @property
    def usedAndTotalMinutesDesc(self) -> str:
        return (f"{self.usedMinutes}/{self.totalMinutes}分钟"
                if self.status == ChargingStation.Status.USING else "")

    @property
    def remainingTimeDesc(self) -> str:
        return (f"{self.remainingMinutes // 60}小时{self.remainingMinutes % 60}分钟"
                if self.status == ChargingStation.Status.USING else "")

    @property
    def endTimeDesc(self) -> str:
        return (getDatetime("%Y-%m-%d %H:%M", +8, self.remainingMinutes)
                if self.status == ChargingStation.Status.USING else "")

    @property
    def note(self) -> str:
        return (self.status.value
                if self.status == ChargingStation.Status.UNAVAILABLE else "")


def extractDigit(s: str) -> int:
    try:
        return int("".join(list(filter(str.isdigit, s))))
    except:
        return 0


def getDatetime(fmt="%Y-%m-%d %H:%M:%S", timezone=8, deltaMinutes=0):
    import datetime, time
    return ((datetime.datetime.utcfromtimestamp(time.time())
             + datetime.timedelta(hours=timezone, minutes=deltaMinutes))
            .strftime(fmt))


async def checkAlive(session: ClientSession) -> bool:
    async with session.get("https://api.issks.com/issksapi/V2/ec/userInfo.shtml") as resp:
        text = await resp.text()
        return text.strip() != ""


async def fetchOutletList(session: ClientSession, station: dict) -> list[dict]:
    async with session.get("https://api.issks.com/issksapi/V2/ec/chargingList.shtml",
                           params=dict(stationId=station["iStationId"])) as resp:
        assert resp.status == 200
        return (await resp.json())["list"]


async def fetchOutletDetail(session: ClientSession, station: dict, outlet: dict) -> ChargingStation:
    async with session.get(f"https://api.issks.com/issksapi/V2/ec/charging/{outlet['vOutletNo']}.shtml") as resp:
        assert resp.status == 200
        text = await resp.text()
        chargingStation = ChargingStation(station["vStationName"], outlet["vOutletName"], station["area"])
        if "设备维护中" in text or outlet["status"] == 3:  # status=3 时页面提示该插座安全隐患不可用
            chargingStation.status = ChargingStation.Status.UNAVAILABLE
        elif (soup := BeautifulSoup(text, "lxml")).select_one(".state_item"):
            chargingStation.status = ChargingStation.Status.USING
            chargingStation.power = extractDigit(soup.select_one(".state_item:nth-child(1) p").text)  # 瓦
            chargingStation.usedMinutes = extractDigit(soup.select_one(".state_item:nth-child(2) p").text)  # 分钟
            totalMinutes = extractDigit(soup.select(".state_item:nth-child(1) span")[-1].text) * 60  # 分钟
            chargingStation.totalMinutes = totalMinutes if totalMinutes != 0 else 600
        return chargingStation


async def main():
    session = aiohttp.ClientSession(trust_env=True)
    session.headers.update({
        "User-Agent": "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36 NetType/WIFI MicroMessenger/7.0.20.1781(0x6700143B) WindowsWechat(0x6307061d)",
        "Host": "api.issks.com"
    })
    session.cookie_jar.update_cookies(dict(
        JSESSIONID=os.getenv("JSESSIONID"),
    ))

    # 验证登录有效
    if not await checkAlive(session):
        print("Cookie失效")
        exit(-1)

    tasks = []
    for station in stations:
        tasks.append(asyncio.ensure_future(fetchOutletList(session, station)))
    outletLists = await asyncio.gather(*tasks)

    tasks = []
    for station, outlets in zip(stations, outletLists):
        for outlet in outlets:
            tasks.append(asyncio.ensure_future(fetchOutletDetail(session, station, outlet)))
    results = await asyncio.gather(*tasks)

    await session.close()

    results = sorted(results, key=lambda x: x.remainingMinutes)

    html = """
    <style>
    table, th, td {
      border: 1px solid;
    }
    </style>
    """ + f"""
    <h1>更新时间：{getDatetime(fmt="%Y-%m-%d %H:%M")}</h1>
    """
    htmlTableHeader = ("充电桩", "插座号", "充电时长", "剩余时长", "结束时间", "备注")

    for area in list(OrderedDict.fromkeys([s["area"] for s in stations])):
        html += f"<h3>{area}</h3>"
        html += tabulate(list(map(lambda x: [x.stationName, x.outletName, x.usedAndTotalMinutesDesc,
                                             x.remainingTimeDesc, x.endTimeDesc, x.note],
                                  filter(lambda x: x.area == area, results))),
                         htmlTableHeader, tablefmt="html")

    if not os.path.exists("build"):
        os.mkdir("build")
    open("build/index.html", mode="w", encoding="utf-8").write(html)


asyncio.run(main())

File: src/main/kotlin/Spider.kt
import com.fasterxml.jackson.databind.node.ArrayNode
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.html.body
import kotlinx.html.h1
import kotlinx.html.h3
import kotlinx.html.head
import kotlinx.html.html
import kotlinx.html.p
import kotlinx.html.stream.createHTML
import kotlinx.html.style
import kotlinx.html.table
import kotlinx.html.tbody
import kotlinx.html.td
import kotlinx.html.th
import kotlinx.html.thead
import kotlinx.html.title
import kotlinx.html.tr
import model.Outlet
import model.Station
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import util.JsonUtils
import util.OkHttpUtils
import util.TimeUtils
import util.syncGetJson
import java.io.File
import java.net.InetSocketAddress
import java.net.Proxy
import java.net.URI
import java.util.concurrent.CopyOnWriteArrayList

class Spider {

    private val token = System.getenv("token")
        ?: File("token.json").readTextOrNull()?.trim()
        ?: ""

    private val proxy = System.getenv("HTTP_PROXY")
        ?.let { URI(it) }
        ?.let { Proxy(Proxy.Type.HTTP, InetSocketAddress(it.host, it.port)) }
        ?: Proxy.NO_PROXY

    private val headerInterceptor = Interceptor { chain ->
        val originalRequest = chain.request()

        val modifiedRequest = originalRequest.newBuilder()
            .header(
                "User-Agent",
                "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36 NetType/WIFI MicroMessenger/7.0.20.1781(0x6700143B) WindowsWechat(0x6307061d)",
            )
            .addHeader("token", token)
            .build()

        chain.proceed(modifiedRequest)
    }

    private val client: OkHttpClient = OkHttpUtils.newBuilder(true)
        .addInterceptor(headerInterceptor)
        .proxy(proxy)
        .build()

    private fun File.readTextOrNull(charset: java.nio.charset.Charset = Charsets.UTF_8): String? {
        return try {
            this.readText(charset)
        } catch (e: Exception) {
            null
        }
    }

    fun checkAlive(): Boolean {
        return try {
            val json = client.syncGetJson(
                "https://wemp.issks.com/recharge/v1/registerCard/checkSign"
            )
            json.get("success").asBoolean()
        } catch (e: Exception) {
            false
        }
    }

    fun fetchOutletNo(station: Station): List<String> {
        val json = client.syncGetJson(
            "https://wemp.issks.com/charge/v1/outlet/station/outlets/${station.id}"
        )
        return (json.get("data") as ArrayNode).map {
            it.get("outletNo").asText()
        }
    }

    private fun String.extractDigit(): Int = try {
        this.filter { it.isDigit() }.toInt()
    } catch (ignored: Exception) {
        0
    }

    fun fetchOutletDetail(outletNo: String): Outlet {
        val json = client.syncGetJson(
            "https://wemp.issks.com/charge/v1/charging/outlet/${outletNo}"
        )
        val data = json["data"]
        val stationData = data["station"]
        val outletData = data["outlet"]

        val stationId = stationData["iStationId"].asInt()
        val outletName = outletData["vOutletName"].asText()
        val usedMin = data["usedmin"].asInt()
        val remainingMin = data["restmin"].asInt()
        val power = json["powerFee"]?.get("billingPower")?.asText()?.extractDigit() ?: 0
        val status = if (data["station"]["hardWareState"].asInt() != 1) {
            Outlet.Status.UNAVAILABLE
        } else if (data["outlet"]["iCurrentChargingRecordId"].asInt() != 0) {
            Outlet.Status.USING
        } else if (data["outlet"]["iErrorCount"].asInt() > 0){
            Outlet.Status.UNAVAILABLE
        } else {
            Outlet.Status.AVAILABLE
        }
        val totalMin = when (status) {
            Outlet.Status.AVAILABLE -> 0
            Outlet.Status.USING -> if (remainingMin > 0) usedMin + remainingMin else 999
            Outlet.Status.UNAVAILABLE -> 1000
        }

        return Outlet(
            outletName = outletName,
            stationId = stationId,
            status = status,
            power = power,
            usedMinutes = usedMin,
            totalMinutes = totalMin,
        )
    }

}

fun main() {

    val stations = JsonUtils.toJson(File("stations.json").readText())
        ?.map {
            Station(
                it.get("id").asInt(),
                it.get("name").asText(),
                it.get("area").asText(),
            )
        }
        ?: error("Read stations.json failed.")

    val stationId2Station = stations.associateBy { it.id }

    val stationId2OutletNos = mutableMapOf<String, List<String>>()
    JsonUtils.toJson(File("outlets.json").readText())?.fields()?.forEach {
        stationId2OutletNos[it.key] = it.value.map { it.asText() }
    }

    val spider = Spider()
    if (spider.checkAlive()) {
        runBlocking {
            for (station in stations) {
                launch(Dispatchers.IO) {
                    val outletNo = spider.fetchOutletNo(station)
                    stationId2OutletNos[station.id.toString()] = outletNo
                }
            }
        }
        // write to disk
        JsonUtils.toJsonString(stationId2OutletNos)?.let {
            File("outlets.json").writeText(it)
        }
    }
    if (stationId2OutletNos.isEmpty()) {
        return
    }

    val results: MutableList<Outlet> = CopyOnWriteArrayList()
    runBlocking{
        for (station in stations) {
            stationId2OutletNos[station.id.toString()]?.forEach { outletNo ->
                launch(Dispatchers.IO) {
                    val chargingStation = spider.fetchOutletDetail(outletNo)
                    results.add(chargingStation)
                }
            }
        }
    }

    results.sortBy { it.remainingMinutes }

    val html = createHTML().html {
        head {
            title("南哪儿充不了电")
        }
        body {
            style {
                +"table, th, td { border: 1px solid; }"
            }
            h1 { +"更新时间：${TimeUtils.getDatetime()}" }
            p { +"status:${if (results.isNotEmpty()) "up" else "down"}" }
            for (area in stations.map { it.area }.distinct()) {
                h3 { +area }
                table {
                    thead {
                        tr {
                            th { +"充电桩" }
                            th { +"插座号" }
                            th { +"充电时长" }
                            th { +"剩余时长" }
                            th { +"结束时间" }
                            th { +"备注" }
                        }
                    }
                    tbody {
                        for (result in results) {
                            val station = stationId2Station[result.stationId]
                            if (station?.area == area) {
                                tr {
                                    td { +station.name }
                                    td { +result.outletName }
                                    td { +result.usedAndTotalMinutesDesc }
                                    td { +result.remainingTimeDesc }
                                    td { +result.endTimeDesc }
                                    td { +result.note }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    File("build/html/index.html").let {
        it.parentFile.mkdirs()
        it.writeText(html)
    }
}
